### ISSUE TITLE
Improve API for "snaking" grid_scan

### ISSUE THREAD

**Issue by danielballan** [bluesky/bluesky#1189]
The API for optionally "snaking" certain axes of a `grid_scan` [as explained here](https://blueskyproject.io/bluesky/tutorial.html#scan-multiple-motors-in-a-grid) is a bit confusing.

Current API:
```python
grid_scan(detectors,
    motor1, start1, stop1, num_steps1,
    motor2, start2, stop2, num_steps2, snaked_or_not2,
    motor3, start3, stop3, num_steps3, snaked_or_not3,
    ...)
```
The fact that the first axis does not take a boolean but all subsequent axes must have a boolean trips people up, and it forces users who don't care about the snaking feature to pause and understand it.

A new API has been proposed:

```python
grid_scan(detectors,
    motor1, start1, stop1, num_steps1,
    motor2, start2, stop2, num_steps2,
    motor3, start3, stop3, num_steps3,
    ...,
    snaked=SNAKED)
# where snaked is an optional, keyword-only argument, and SNAKED can be:
# False (default) -- Do not snake any axes.
# True -- Snake all axes
# tuple of booleans -- Snake some specific axes
```

I think this new usage is better in every way; I wish we had thought of it in the first place! Fortunately, it is possible to change `grid_scan` to support both the old and new usages by duck-typing the `args` and figuring out where the motors are. I propose to do that, advertise the new usage, and demote the old usage to footnote in the docstring.

Alternatively, we could write a separate function, but then we lose the name `grid_scan` and need something like `new_grid_scan` going forward. This is why I favor updating `grid_scan` in a backward-compatible way.

**Comment by ambarb** [comment]
This makes a lot of sense to change the documentation and how the function arguments work.  I would recommend that the doc string contain an example on how to use the tuple of booleans.  I naively assume if I want to snake motor1 and motor3 for a 4 motor scan, then `snaked=(True,False,True, False)` respective to list of motor argument . If this is correct, then I should get python points. I am  just trying to say that I am not that all of our users know what a tuple is.

**Comment by danielballan** [comment]
Python points for you! :-D

We will definitely document this with examples.

**Comment by danielballan** [comment]
Any concerns from @prjemian, @teddyrendahl, @ZLLentz?

**Comment by danielballan** [comment]
One finer point to work out:

If there are N motors, the first motor _can't_ be snaked. Do we accept N-1 settings in the tuple, as in:

```python
grid_scan(detectors,
    motor1, start1, stop1, num_steps1,
    motor2, start2, stop2, num_steps2,
    motor3, start3, stop3, num_steps3,
    snaked=(True, False))  # refers motor2, motor3 respectively
```

or

```python
grid_scan(detectors,
    motor1, start1, stop1, num_steps1,
    motor2, start2, stop2, num_steps2,
    motor3, start3, stop3, num_steps3,
    snaked=(False, True, False))  # refers motor1, motor2, motor3 respectively
```

and raise if the first item in the tuple is anything but `False`. I submit that the latter will likely to be less confusing to use. The "off by one-ness" of this throws people off.

**Comment by CJ-Wright** [comment]
I guess we could support both and figure out which one is the right one by inspection of the length.

**Comment by danielballan** [comment]
I think I would not like that because it makes wrapping your brain around the complete "rules" for how to use this and how it behaves even more complex.

**Comment by teddyrendahl** [comment]
I do see this as an API improvement. I am somewhat against having the `N` sized tuple where the first argument is always`False`.  

Spitballing here, any credence to passing in the motors you want snaked instead? 

```python
grid_scan(detectors,
    motor1, start1, stop1, num_steps1,
    motor2, start2, stop2, num_steps2,
    motor3, start3, stop3, num_steps3,
    snaked=(motor2, motor3)
```

**Comment by awalter-bnl** [comment]
+1 for @teddyrendahl suggestion. This is highly intuitive. Maybe “all” for all axes instead of true and None for None?

**Comment by CJ-Wright** [comment]
I would rather a bool be used for all or nothing decisions (with `None` being a proxy for False). String sentinels can get messy. I think a signature with `snake=None|True|(motor2, motor3)` could be interpreted as "do I want to snake?" `True` I would like to snake all the things.

**Comment by prjemian** [comment]
IOM, I prefer the original API, not the proposed one.  Separating the `snake` kwargs into a tuple disconnects them from use.  The parameter list is of monumental length already.  Seems like this is a change that adds an additional way to describe the same thing.

**Comment by prjemian** [comment]
IOM == IMO

**Comment by mrakitin** [comment]
What if a motor object knows about whether to "snake" or not? Just a thought...

**Comment by CJ-Wright** [comment]
How would it know if it needed to snake?

**Comment by prjemian** [comment]
That sounds like a situational decision.  If there is a change, Teddy's [proposal](https://github.com/bluesky/bluesky/issues/1189#issuecomment-492824918) makes sense to me.  And it resolves one of the original problems where new users often have problems learning how to use this plan.

**Comment by mrakitin** [comment]
> How would it know if it needed to snake?

False by default, configurable with setting an atrribute before running or within a plan.

**Comment by awalter-bnl** [comment]
> I would rather a bool be used for all or nothing decisions (with `None` being a proxy for False). String sentinels can get messy. I think a signature with `snake=None|True|(motor2, motor3)` could be interpreted as "do I want to snake?" `True` I would like to snake all the things.

I think for the average user it is easier to think “which motors do I want to snake” None, all or a list of specific ones.

If the question is rephrased as “do I want to  the answer should be “yes”, “no” or “motors in this list”
And this is less intuitive, as is the use of True and False.

**Comment by awalter-bnl** [comment]
> What if a motor object knows about whether to "snake" or not? Just a thought...

This is moving the logic to a place that is not so obvious for an average user. Keeping it in the plan API I think is a better option.

**Comment by danielballan** [comment]
> Separating the snake kwargs into a tuple disconnects them from use. 

I had this thought too, @prjemian. Indeed, there is _some_ advantage to keeping each snake parameter right next to the group that it belongs to. But overall maybe the pros of this new proposal outweigh the cons, particularly for new users.

It looks like we are all at least content with @teddyrendahl's refinement on this proposal, judging from the :+1:s. We can continue to support the original way indefinitely for situations/users that find more readability there.

Any objections to proceeding in that way?

**Comment by awalter-bnl** [comment]
Which of the True/False, “all”/None, “Yes”/“No” options are we suggesting using?

**Comment by teddyrendahl** [comment]
I vote for boolean for all or nothing behavior. Generally don't like having to remember special string names especially for binary options.

**Comment by danielballan** [comment]
Yeah, I agree that anything but normal boolean would be very unusual. I can't think of an example of a scipy library that uses "Yes"/"No" or "all"/`None` when `True`/`False` would work.

> What is going to be the default behavior since this was a mandatory argument before?

My original suggestion was `False` by default because that seems most likely to at least _work_ (if not necessary the fastest). Some motors can't scan in reverse, I'm told.

**Comment by prjemian** [comment]
Some motors do not scan well in both directions, usually due to gear backlash.  The EPICS motor record has code to account for this but with real hardware, can only do so much.  The correction involves target overshoot and then approach from the other direction.  Data are more reproducible with such axes when scanning in the direction where backlash correction is not needed.

**Comment by prjemian** [comment]
I'm not clear with https://github.com/bluesky/bluesky/issues/1189#issuecomment-492867947.  Is that `snaked` can take any of these values?

* `None` (default) - no axes are snaked
* `True` - all axes but the first are snaked
* list of axes names that are snaked

**Comment by danielballan** [comment]
Yes. I had in mind `False` as the default value, but anything "false-y" would have the same outcome, so `snaked=None` would have the same result as `snaked=False`.

**Comment by danielballan** [comment]
And thanks for the explanation @prjemian. This is why I think snaking should be an opt-in feature.

**Comment by prjemian** [comment]
That's better.

**Comment by prjemian** [comment]
Changed my table so it is clear.

**Comment by awalter-bnl** [comment]
Ok so given that I am in the minority by 4:1 I am ok with moving forward with this suggestion even though I disagree on the no axes and all axes values.

**Comment by danielballan** [comment]
This approach has been applied to `list_grid_scan` by @awalter-bnl in #1190. It remains to apply it to `grid_scan` in a backward-compatible way. Per discussion above, it looks like we agreed that this change is worth making because (quoting @prjemian) "it resolves one of the original problems where new users often have problems learning how to use this plan." We can support the original usage as well, indefinitely.

I will assign the work of changing `grid_scan` to myself. Once that's done, this issue can be closed.

**Comment by danielballan** [comment]
Some thought will need to be given to how to sniff out the types of the parameters. The start and stop parameters can in principle be anything (not necessary just a number) so I think the best bet is find where the motors are by using something like

```py
In [21]: class Movable(metaclass=abc.ABCMeta): 
    ...:     "Abstract base class for objects that satisfy the bluesky 'movable' interface." 
    ...:     @classmethod 
    ...:     def __subclasshook__(cls, C): 
    ...:         return bool(hasattr(C, 'name') 
    ...:                     and hasattr(C, 'set') 
    ...:                     and ...) 
    ...:                      
    ...:                                                                                                                                                                                      

In [22]: from ophyd.sim import motor                                                                                                                                                          

In [23]: isinstance(motor, Movable)                                                                                                                                                           
Out[23]: True

In [24]: isinstance(3, Movable)                                                                                                                                                               
Out[24]: False
```

where `Movable` is intended to be used not as a _base class_ but as a duck-type checker like `collections.abc.Iterable`.

**Comment by danielballan** [comment]
... where `...` above is a placeholder for more methods that should be checked for. See https://blueskyproject.io/bluesky/hardware.html#readable-device.

**Comment by ksunden** [comment]
I believe this has been resolved by #1334, or am I missing some additional changes desired?

**Comment by danielballan** [comment]
Thanks @ksunden. You are correct, I believe. Closing.
